### ISSUE TITLE
Question about the weight for correction in the importance sampling

### ISSUE THREAD

**Issue by miracle24** [awarebayes/RecNN#7]
Hello. Thanks for your great work, from which I learned a lot about reinforcement learning. I am confused about the computation of the correction weight in the importance sampling. 

According to the paper "Top-K Off-Policy Correction for a REINFORCE Recommender System", the correction weight is  ![](http://latex.codecogs.com/gif.latex?\\frac{\pi(a_t|s_t)}{\beta(a_t|s_t)}), in which, I think, ![](http://latex.codecogs.com/gif.latex?\\\a_t) is an action sampled from the behavior policy, i.e. ![](http://latex.codecogs.com/gif.latex?\\beta), and thus the reward of a sequence ![](http://latex.codecogs.com/gif.latex?\\tau) is corrected by dividing the likelihood of action ![](http://latex.codecogs.com/gif.latex?\\\a_t) given ![](http://latex.codecogs.com/gif.latex?\\\s_t) in the updated policy. i.e. ![](http://latex.codecogs.com/gif.latex?\\pi) by the likelihood of **the same action** ![](http://latex.codecogs.com/gif.latex?\\\a_t) given ![](http://latex.codecogs.com/gif.latex?\\\s_t) in ![](http://latex.codecogs.com/gif.latex?\\beta). Noted ![](http://latex.codecogs.com/gif.latex?\\\a_t) will be input into both ![](http://latex.codecogs.com/gif.latex?\\pi) and ![](http://latex.codecogs.com/gif.latex?\\beta). However, in your implementation, I found that actions are not the same for ![](http://latex.codecogs.com/gif.latex?\\pi) and ![](http://latex.codecogs.com/gif.latex?\\beta), seeing the function "pi_beta_sample" in [here](https://github.com/awarebayes/RecNN/blob/master/recnn/nn/models.py). 

Am I wrong about this?

Thanks!

**Comment by awarebayes** [comment]
# Ok, let's try to figure it out.

Note: 
- historical a_t is the action from the dataset that you are referring to.
- beta a_t is an action that was generated by the beta network. Should be the same as historical a_t
- alpha a_t is an action by the actor that learns it using the reinforce algorithm    
    
## 1. What if we use a_t?
```
    def pi_beta_sample(self, state, beta, action, **kwargs):
       # action is a_t       
 
        # 1. obtain probabilities
        # note: detach is to block gradient
        beta_probs = beta(state.detach(), action=action)
        pi_probs = self.forward(state)

        # 2. probabilities -> categorical distribution
        beta_categorical = Categorical(beta_probs)
        pi_categorical = Categorical(pi_probs)

        # 4. Ignore all the samples, learn from the historical action
        pi_log_prob = pi_categorical.log_prob(action)
        beta_log_prob = beta_categorical.log_prob(action)

        return pi_log_prob, beta_log_prob, pi_probs
```

In that case, both Alpha and Beta will just be learning the historical action from the dataset, no matter the reward or the recommended action. In the REINFORCE algorithm we are required to learn the Actor action, not the historical one:
```
    # REINFORCE actor without beta
    def _select_action(self, state, **kwargs):
        pi_probs = self.forward(state)
        pi_categorical = Categorical(pi_probs)
        pi_action = pi_categorical.sample()
        self.saved_log_probs.append(pi_categorical.log_prob(pi_action))
        return probs
```

So this is important to learn action that was sampled from the Actor probabilities:

```
        # 3. sample actions
        beta_action = beta_categorical.sample()
        pi_action = pi_categorical.sample()

        # 4. calculate stuff we need
        pi_log_prob = pi_categorical.log_prob(pi_action)
        beta_log_prob = beta_categorical.log_prob(beta_action)
```

It will not learn otherwise. This is the algorithm itself described in the paper, and I mainly copied it from the official PyTorch examples:  https://github.com/pytorch/examples/blob/master/reinforcement_learning/reinforce.py

TL;DR; REINFORCE needs log prob pi sampling, this is the nature of the algorithm.

So, we understand that at least for Alpha this is justified. Now, let's consider the beta part:

## 2. Historical A_t vs Beta Generated A_t
As you have said, beta needs to use the historical action for correction. But since that beta tries to learn that action, it is practically the same as the historical a_t. I think the main purpose of the beta network is to predict historical actions based on the encoded state. 

But you also can substitute beta action with the historical one. I don't know what happens, but it should be the same.

**Comment by awarebayes** [comment]
But I will consider changing it as you said. Probably that will be a better solution

**Comment by awarebayes** [comment]
Now that I am looking at it, I think this will be a more accurate solution :

```
def pi_beta_sample(self, state, beta, action, **kwargs):
 # 1. obtain probabilities
 # note: detach is to block gradient
beta_probs = beta(state.detach(), action=action)
pi_probs = self.forward(state) 

# 2. probabilities -> categorical 
distribution beta_categorical = Categorical(beta_probs) 
pi_categorical = Categorical(pi_probs) 
# 3. sample actions 
beta_action = beta_categorical.sample() 
pi_action = pi_categorical.sample() 
# 4. calculate stuff we need 
pi_log_prob = pi_categorical.log_prob(action)
beta_log_prob = beta_categorical.log_prob(action) 
return pi_log_prob, beta_log_prob, pi_probs
```

I will update the repo and the article soon. 
Thanks for helping me spot that

**Comment by awarebayes** [comment]
Alright, about to release a commit to this. 

The solution is simple: substitute the historical action with the one generated by the beta network. Why do we need it? Because we need the probabilities.
P.S. Historical a_t is just a number, movie_id if you like. 

Can we do that? Since the entire purpose of the beta is to emulate the historical action, it can give us the probabilities. As well that historical action you were talking about.

I have addded a parameter called 'action_source' and so far I came up with:
```
class DiscreteActor(nn.Module):
    def __init__(self, input_dim, action_dim, hidden_size, init_w=0):
        super(DiscreteActor, self).__init__()
        ....
        # What's action source? See this issue: https://github.com/awarebayes/RecNN/issues/7
        # by default {pi: pi, beta: beta}
        # you can change it to be like {pi: beta, beta: beta} as miracle24 suggested
        self.action_source = {'pi': 'pi', 'beta': 'beta'}

    def pi_beta_sample(self, state, beta, action, **kwargs):
        ....

        # 3. sample the actions
        # See this issue: https://github.com/awarebayes/RecNN/issues/7
        # usually it works like:
        # pi_action = pi_categorical.sample(); beta_action = beta_categorical.sample();
        # but changing the action_source to {pi: beta, beta: beta} can be configured to be:
        # pi_action = beta_categorical.sample(); beta_action = beta_categorical.sample();
        available_actions = {'pi': pi_categorical.sample(), 'beta': beta_categorical.sample()}
        pi_action = available_actions[self.action_source['pi']]
        beta_action = available_actions[self.action_source['beta']]

policy_net = recnn.nn.DiscreteActor(1290, num_items, 2048).to(cuda)
# as miracle24 has suggested https://github.com/awarebayes/RecNN/issues/7
# you can enable this to be more like the paper authors meant it to
# policy_net.action_source = {'pi': 'beta', 'beta': 'beta'}
```

**Comment by awarebayes** [comment]
I have released a commit to this, [you can look at it here](https://github.com/awarebayes/RecNN/commit/0945ec0db294a8828dad0c1983c55d8b955895be)

Write me back if I can close the issue

**Comment by miracle24** [comment]
Actually, I am not sure which one is correct or better. Maybe I can try both in my task. Thanks a lot.
